src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs

```diff
@@ -36,202 +79,565 @@ public SafeX509ChainHandle SafeHandle
-                            cert,
-                            store,
-                            revocationMode,
-                            verificationTime,
-                            ref remainingDownloadTime);
+                        break;
+                    }
 
-                        // If we only wanted the end-entity certificate CRL then don't look up
-                        // any more of them.
-                        lookupCrl = revocationFlag != X509RevocationFlag.EndCertificateOnly;
+                    if (downloadedCerts == null)
+                    {
+                        downloadedCerts = new List<X509Certificate2>();
                     }
+
+                    AddToStackAndUpRef(downloaded.Handle, _untrustedLookup);
+                    downloadedCerts.Add(downloaded);
+                    
+                    Interop.Crypto.X509StoreCtxRebuildChain(storeCtx);
+                    statusCode = Interop.Crypto.X509StoreCtxGetError(storeCtx);
                 }
+            }
 
-                if (revocationMode != X509RevocationMode.NoCheck)
+            if (statusCode == Interop.Crypto.X509VerifyStatusCode.X509_V_OK && downloadedCerts != null)
+            {
+                using (SafeX509StackHandle chainStack = Interop.Crypto.X509StoreCtxGetChain(_storeCtx))
                 {
-                    if (!Interop.Crypto.X509StoreSetRevocationFlag(store, revocationFlag))
+                    int chainSize = Interop.Crypto.GetX509StackFieldCount(chainStack);
+                    Span<IntPtr> tempChain = stackalloc IntPtr[DefaultChainCapacity];
+                    IntPtr[] tempChainRent = null;
+
+                    if (chainSize < tempChain.Length)
                     {
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                        tempChain = tempChain.Slice(chainSize);
+                    }
+                    else
+                    {
+                        tempChainRent = ArrayPool<IntPtr>.Shared.Rent(chainSize);
+                        tempChain = tempChainRent.AsSpan(0, chainSize);
                     }
-                }
 
-                foreach (X509Certificate2 trustedCert in systemTrusted)
-                {
-                    OpenSslX509CertificateReader pal = (OpenSslX509CertificateReader)trustedCert.Pal;
+                    for (int i = 0; i < chainSize; i++)
+                    {
+                        tempChain[i] = Interop.Crypto.GetX509StackField(chainStack, i);
+                    }
 
-                    if (!Interop.Crypto.X509StoreAddCert(store, pal.SafeHandle))
+                    for (int i = downloadedCerts.Count - 1; i >= 0; i--)
                     {
-                        throw Interop.Crypto.CreateOpenSslCryptographicException();
+                        X509Certificate2 downloadedCert = downloadedCerts[i];
+
+                        if (!tempChain.Contains(downloadedCert.Handle))
```

Comment:
The O(m*n) here is ok because m and n are always small?

Reply:
Yeah, `downloaded` is usually maximally 2, and tempChain is usually maximally 3; so it didn't seem worth a temporary hashset.